src/qibo/transpiler/unroller.py

```diff
@@ -0,0 +1,460 @@
+        """Register a decomposition for a gate."""
+        self.decompositions[gate] = decomposition
+
+    def count_2q(self, gate):
+        """Count the number of two-qubit gates in the decomposition of the given gate."""
+        if gate.parameters:
+            decomposition = self.decompositions[gate.__class__](gate)
+        else:
+            decomposition = self.decompositions[gate.__class__]
+        return len(tuple(g for g in decomposition if len(g.qubits) > 1))
+
+    def count_1q(self, gate):
+        """Count the number of single qubit gates in the decomposition of the given gate."""
+        if gate.parameters:
+            decomposition = self.decompositions[gate.__class__](gate)
+        else:
+            decomposition = self.decompositions[gate.__class__]
+        return len(tuple(g for g in decomposition if len(g.qubits) == 1))
+
+    def __call__(self, gate):
+        """Decompose a gate."""
+        decomposition = self.decompositions[gate.__class__]
+        if callable(decomposition):
+            decomposition = decomposition(gate)
+        return [
+            g.on_qubits({i: q for i, q in enumerate(gate.qubits)})
+            for g in decomposition
+        ]
+
+
+onequbit_dec = GateDecompositions()
+onequbit_dec.add(gates.H, [gates.U3(0, 7 * np.pi / 2, np.pi, 0)])
+onequbit_dec.add(gates.X, [gates.U3(0, np.pi, 0, np.pi)])
+onequbit_dec.add(gates.Y, [gates.U3(0, np.pi, 0, 0)])
+# apply virtually by changing ``phase`` instead of using pulses
+onequbit_dec.add(gates.Z, [gates.Z(0)])
+onequbit_dec.add(gates.S, [gates.RZ(0, np.pi / 2)])
+onequbit_dec.add(gates.SDG, [gates.RZ(0, -np.pi / 2)])
+onequbit_dec.add(gates.T, [gates.RZ(0, np.pi / 4)])
+onequbit_dec.add(gates.TDG, [gates.RZ(0, -np.pi / 4)])
+onequbit_dec.add(
+    gates.RX, lambda gate: [gates.U3(0, gate.parameters[0], -np.pi / 2, np.pi / 2)]
+)
+onequbit_dec.add(gates.RY, lambda gate: [gates.U3(0, gate.parameters[0], 0, 0)])
+# apply virtually by changing ``phase`` instead of using pulses
+onequbit_dec.add(gates.RZ, lambda gate: [gates.RZ(0, gate.parameters[0])])
+# apply virtually by changing ``phase`` instead of using pulses
+onequbit_dec.add(gates.GPI2, lambda gate: [gates.GPI2(0, gate.parameters[0])])
+# implemented as single RX90 pulse
+onequbit_dec.add(gates.U1, lambda gate: [gates.RZ(0, gate.parameters[0])])
+onequbit_dec.add(
+    gates.U2,
+    lambda gate: [gates.U3(0, np.pi / 2, gate.parameters[0], gate.parameters[1])],
+)
+onequbit_dec.add(
+    gates.U3,
+    lambda gate: [
+        gates.U3(0, gate.parameters[0], gate.parameters[1], gate.parameters[2])
+    ],
+)
+onequbit_dec.add(
+    gates.Unitary,
+    lambda gate: [gates.U3(0, *u3_decomposition(gate.parameters[0]))],
+)
+onequbit_dec.add(
+    gates.FusedGate,
+    lambda gate: [gates.U3(0, *u3_decomposition(gate.matrix(backend)))],
+)
+
+# register the iSWAP decompositions
+iswap_dec = GateDecompositions()
+iswap_dec.add(
+    gates.CNOT,
+    [
+        gates.U3(0, 7 * np.pi / 2, np.pi, 0),
```

Comment:
Isn't $7\\pi / 2$ outside of the domain of $\\theta \\in [0, 2\\pi]$? In any case it's periodic, then maybe this could be either $3\\pi/2$.